### PR TITLE
Fixes #2057

### LINKED ISSUE #2057: My server connects is not giving error on the console
```
Genisys Crash Dump Tue Oct 11 05:14:35 EDT 2016

Error: Argument 1 passed to pocketmine\permission\BanEntry::setCreated() must be an instance of DateTime, boolean given, called in phar:///home/multicraft/jar/Genisys.phar/src/pocketmine/permission/BanEntry.php on line 112
File: /src/pocketmine/permission/BanEntry
Line: 48
Type: notice

Code:
[39] 
[40] 	public function getName() : string{
[41] 		return $this->name;
[42] 	}
[43] 
[44] 	public function getCreated(){
[45] 		return $this->creationDate;
[46] 	}
[47] 
[48] 	public function setCreated(\DateTime $date){
[49] 		$this->creationDate = $date;
[50] 	}
[51] 
[52] 	public function getSource(){
[53] 		return $this->source;
[54] 	}
[55] 
[56] 	public function setSource($source){
[57] 		$this->source = $source;
[58] 	}

Backtrace:
#0 /src/pocketmine/permission/BanList(137): pocketmine\permission\BanEntry::fromString(array Array())
#1 /src/pocketmine/Server(1849): pocketmine\permission\BanList->load(boolean)
#2 /src/pocketmine/PocketMine(476): pocketmine\Server->__construct(pocketmine\CompatibleClassLoader object, pocketmine\utils\MainLogger object, string phar:///home/multicraft/jar/Genisys.phar/, string /home/multicraft/servers/server7346/, string /home/multicraft/servers/server7346/plugins/, string unknown)
#3 (1): require_once(string phar:///home/multicraft/jar/Genisys.phar/src/pocketmine/PocketMine.php)

Genisys version: b816e46335cf145b8eca9c1a24b18ee10e80cae3 [Protocol 84; API 2.0.0]
uname -a: Linux s6.craftandohosting.net 3.14.32-xxxx-grs-ipv6-64 #7 SMP Wed Jan 27 18:05:09 CET 2016 x86_64
PHP version: 7.0.10
Zend version: 3.0.0
OS : Linux, linux

Server uptime: 0 seconds
Number of loaded worlds: 0
Players online: 0/
```

### PR DESCRIPTION
**Pull request by jjplaying**
### Description

Fixes #2057, it replaces | with nothing to avoid bugs in the ban.txt.

### Tests & Reviews

Tested the commands and it works.

### CODE REVIEW

**Comment by dktapps** [comment]
That isn't the only character that causes such a crash.

**Comment by jjplaying** [comment]
Which ones? Someone sad it also crashes with `_`.

**Comment by dktapps** [comment]
This isn't really an effective solution. You should whitelist allowed characters, not blacklist disallowed ones. There are myriad characters this could occur wity.

**Comment by FaigerSYS** [comment]
@jjplaying , false. "_" symbol didn't crash server. Revert this

**Comment by FaigerSYS** [comment]
As @SOF3 said, you can just use `ctype_alnum(str_replace("_", "", $name))` check

**Comment by jjplaying** [comment]
Sounds logical, I'll change this.

**Comment by jjplaying** [comment]
Oh, didn't checked this, too.

**Comment by jjplaying** [comment]
return false; should be okay, because nobody can have those characters in his name.

**Comment by jjplaying** [comment]
Oh shit I can't write on my phone, fix coming soon.

File: src/pocketmine/command/defaults/BanCommand.php [hunk 1]
```diff
@@ -49,9 +49,14 @@ public function execute(CommandSender $sender, $currentAlias, array $args){
 			return false;
 		}
 
-		$name = ctype_alnum(str_replace("_", "", array_shift($args)));
+		$name = str_replace("_", "", array_shift($args));
```
Comment (FaigerSYS) [inline src/pocketmine/command/defaults/BanCommand.php]: You shouldn't replace "_" symbol. You can just remove str_replace here.
`$name = array_shift($args);
if (!ctype_alnum(str_replace("_", "", $name))) {
    $sender->sendMessage("Error message");
    return false;
}`

File: src/pocketmine/command/defaults/BanCommand.php [hunk 2]
```diff
@@ -49,9 +49,14 @@ public function execute(CommandSender $sender, $currentAlias, array $args){
...
 		if(isset($args[0]) and isset($args[1])){
-			$reason = ctype_alnum(str_replace("_", "", $args[0]));
+			$reason = str_replace("_", "", $args[0]);
```
Comment (FaigerSYS) [inline src/pocketmine/command/defaults/BanCommand.php]: There you can just replace this symbol. Or use another filter
`$reason = str_replace("|", "", $args[0]);`
The same must be used (there)[https://github.com/iTXTech/Genisys/blob/master/src/pocketmine/command/defaults/BanCommand.php#L63]

File: src/pocketmine/command/defaults/BanCommand.php (hunk 1 above)
Comment (dktapps) [inline src/pocketmine/command/defaults/BanCommand.php]: @FaigerSYS Go and read the PHP documentation instead of spamming incorrect comments.

**Comment by jjplaying** [comment]
I'll rewrite it tomorrow.

**Comment by dktapps** [comment]
I think ban lists as YAML or some other config format would make more sense, then there's no restriction on what you can and can't do.

**Comment by jjplaying** [comment]
Yeah, that's a very good idea. I'll start writing such a thing today evening when I'm back from school.

**Comment by FaigerSYS** [comment]
@jjplaying , #2073

**Comment by jjplaying** [comment]
Closed

**Comment by FaigerSYS** [comment]
@jjplaying , new PR. #2078

**Comment by jjplaying** [comment]
Looks good, but also closed. We should dtkapps let do this, as he sad, it was his idea.

You should use nested YAML and not the old format in YAML files, that's better.

**Comment by ishitatsuyuki** [comment]
Some principles:
Keep it simple, stupid
You're not gonna need it

So, only fixing the bug should be enough. Anyway, I'm gonna reviewing it soon.

**Comment by SOF3** [comment]
The output name can contain an underscore...

**Comment by ishitatsuyuki** [comment]
@dktapps if you are free, please give a review.

**Comment by dktapps** [comment]
This doesn't actually fix the initial issue, which is that an invalid banned-players.txt would cause a crash. I fixed that in https://github.com/iTXTech/Genisys/commit/553f2a5a9e75ff2aa1bc3d7d7096323b3f71cc35

This also isn't foolproof. The main issue here is the separator |. Potential issues can also arise from putting a | in the reason, for example.

Finally, I think such a name validation should be a utils static function for consistency throughout the software.

**Pull request closed without merging**